java/ql/src/semmle/code/java/dataflow/FlowSteps.qll

```diff
@@ -69,3 +69,59 @@ private class StringTaintPreservingCallable extends TaintPreservingCallable {
     this.hasName(["format", "formatted", "join"]) and arg = [0 .. getNumberOfParameters()]
   }
 }
+
+private class NumberTaintPreservingCallable extends TaintPreservingCallable {
+  int argument;
+
+  NumberTaintPreservingCallable() {
+    this.getDeclaringType().getASupertype*().hasQualifiedName("java.lang", "Number") and
+    (
+      this instanceof Constructor and
+      argument = 0
+      or
+      this.getName().matches(["to%String", "toByteArray", "%Value"]) and
+      argument = -1
+      or
+      this.getName().matches(["parse%", "valueOf%", "to%String", "decode"]) and
+      argument = 0
+    )
+  }
+
+  override predicate returnsTaintFrom(int arg) { arg = argument }
+}
+
+private class StringBuilderTaintPreservingCallable extends TaintPreservingCallable {
+  StringBuilderTaintPreservingCallable() {
+    exists(Class c | c = this.getDeclaringType().getASourceSupertype*() |
+      (
+        c.hasQualifiedName("java.lang", "StringBuilder") or
+        c.hasQualifiedName("java.lang", "StringBuffer") or
+        c.hasQualifiedName("java.io", "StringWriter")
+      ) and
+      (
+        this.hasName(["append", "insert", "replace", "toString"])
```

Comment:
The `this.getDeclaringType().getASourceSupertype*()` conjunct can be moved here, as it doesn't need to apply to the disjunct below. That in turn frees up the `|` of the `exists`, which can then replace the `and above and thus remove the need for the two sets of parentheses.

Reply:
Maybe include `StringWriter.write`, which transfers taint from arg 0 to the qualifier.